File: backend/routes/bleRouter.js
const express = require('express');
const router = express.Router();
const useBluez = String(process.env.USE_BLUEZ || '0') === '1';
const { listDevices, connect, disconnect } = useBluez
  ? require('../controllers/bleControllerBluez')
  : require('../controllers/bleController');
const { startScanManual, stopScanManual } = require('../services/bleService');
const { startScan: startScanBluez, stopScan: stopScanBluez } = require('../services/bleServiceBluez');

router.get('/devices', listDevices);
router.post('/connect/:id', connect);
router.post('/disconnect/:id', disconnect);
// Fire-and-forget to avoid pending HTTP if BlueZ takes long
router.post('/scan/start', (req, res) => {
  try {
    if (useBluez) { startScanBluez().catch(() => {}); } else { startScanManual(true).catch(() => {}); }
    res.json({ ok: true });
  } catch (e) { res.status(500).json({ ok: false, error: String(e) }); }
});
router.post('/scan/stop', (req, res) => {
  try {
    if (useBluez) { stopScanBluez().catch(() => {}); } else { stopScanManual().catch(() => {}); }
    res.json({ ok: true });
  } catch (e) { res.status(500).json({ ok: false, error: String(e) }); }
});

// Quick status endpoint
router.get('/scan/status', (req, res) => {
  try {
    const list = require('../services/bleService').getDevicesList?.() || [];
    res.json({ ok: true, scanningActive: useBluez ? require('../services/bleServiceBluez').getScanningActive() : require('../services/bleService').getScanningActive(), count: list.length, ts: Date.now() });
  } catch (e) { res.status(500).json({ ok: false, error: String(e) }); }
});

module.exports = router;

File: backend_py/app/routes/ble_router.py
from fastapi import APIRouter
from fastapi.responses import JSONResponse
from typing import Any, Dict
from ..services.ble_service import get_devices_list, start_scan, stop_scan, get_scanning_active, connect_device, disconnect_device

router = APIRouter()

@router.get('/devices')
async def list_devices():
    devices = get_devices_list()
    return JSONResponse({ 'ts': __import__('time').time_ns() // 1_000_000, 'count': len(devices), 'devices': devices })

@router.post('/scan/start')
async def scan_start():
    try:
        await start_scan()
        return JSONResponse({ 'ok': True })
    except Exception as e:
        return JSONResponse({ 'ok': False, 'error': str(e) }, status_code=500)

@router.post('/scan/stop')
async def scan_stop():
    try:
        await stop_scan()
        return JSONResponse({ 'ok': True })
    except Exception as e:
        return JSONResponse({ 'ok': False, 'error': str(e) }, status_code=500)

@router.get('/scan/status')
async def scan_status():
    devices = get_devices_list()
    return JSONResponse({ 'ok': True, 'scanningActive': get_scanning_active(), 'count': len(devices), 'ts': __import__('time').time_ns() // 1_000_000 })

@router.post('/connect/{id}')
async def connect(id: str):
    return await connect_device(id)

@router.post('/disconnect/{id}')
async def disconnect(id: str):
    return await disconnect_device(id)
